### ISSUE TITLE
`all_paths` misses symlinks to directories when not following symlinks

### ISSUE THREAD

**Issue by ncoghlan** [ncoghlan/walkdir#3]
When not following symlinks, symlinks to directories appear in the "subdirs" list, but aren't visited by the underlying walk operation.

This means the current implementation of `all_paths` misses them. The fix likely involves switching to yielding subdirs entries rather than visited directories, but there are a couple of subtleties with that:
- the output order will change such that the parent directory is listed completely before any subdirectories are listed, while the current behaviour is that directory paths are listed immediately prior to the files they contain. Is this OK? (the docs don't currently make any promises about output ordering, and folks that care can use sorted() to get lexical ordering)
- should `dir_paths` be changed as well, or should this be documented as a behaviour difference between the two? (i.e. when not following symlinks `all_paths` may yield more entries than `dir_paths` + `file_paths`, since the latter two will both omit symlinks to directories, while after this change, the former will include them)

**Comment by ncoghlan** [comment]
A couple more points worth noting:
- the current approach yields the initial directory anchoring the walk, so we need to ensure that property is preserved when switching to iterating over the subdirs list
- I initially thought using the new "handle_dirsymlinks_as_files" filter might help, but the problem with that is that we have no way of knowing if the underlying walk is set to follow symlinks or not. If it is, then moving those directories over to the file list will break the recursive descent. That means any `all_paths` changes need to be side-effect free.

**Comment by palaviv** [comment]
I think changing `all_paths` and `dir_paths` should be changed to yield subdirs. This will resolve Issue #4 as well.

I thought about doing something like this:

```
def all_paths(walk_iter):
    """Iterate over both files and directories visited by the underlying walk"""
    dir_entry = next(walk_iter, None)
    if dir_entry is None:
        return
    yield dir_entry[0]
    while dir_entry:
        for fname in dir_entry[2]:
            yield os.path.join(dir_entry[0], fname)
        for subdir in dir_entry[1]:
            yield os.path.join(dir_entry[0], subdir)
        dir_entry = next(walk_iter, None)
```

The problem with this solution is that `min_depth` does not work. I will keep working on this but If we do yield subdirs we should close PR #2 so I will pause with the CR reply.

**Comment by palaviv** [comment]
This issue should be closed after merging #8.
